File: redis/gcra.lua
-- GCRA (Generic Cell Rate Algorithm) rate limiter.
--
-- One atomic EVAL, one Redis round trip per check (constraint 1). The current
-- time is derived from Redis's own TIME so every replica shares one
-- authoritative time base and the caller's clock can never skew the bucket.
--
-- Bucket form of GCRA:
--   bucket capacity = burst
--   one token refills every `interval` ms, where interval = period / rate
--   stored state  = TAT (theoretical arrival time) of the next token, in
--                   epoch milliseconds
--   tokens(t)     = burst - (TAT - t) / interval
--
-- TAT is the only thing stored, and it is a single plain number, so the state
-- is small, orderable, and safe to reset (constraint 6) by simply deleting
-- the key: a missing key is indistinguishable from a fresh bucket. Losing it
-- to a failover likewise just resets to a fresh bucket.
--
-- KEYS[1]  gcra:{client_id}
-- ARGV[1]  burst           bucket capacity in tokens
-- ARGV[2]  rate            tokens per period
-- ARGV[3]  period_seconds  length of one rate period
-- ARGV[4]  cost            tokens consumed by this request (>= 1)
--
-- Returns {allowed, remaining, reset_at_ms, now_ms}:
--   allowed    1/0
--   remaining  floor of tokens in the bucket now, never negative
--   reset_at   on allow: when the bucket is full again (= the new TAT)
--              on deny:  when `cost` tokens are available again (Retry-After)
--   now        the script's own clock (Redis TIME), so callers can compute a
--              Retry-After window on the same clock the bucket was evaluated

local burst  = tonumber(ARGV[1])
local rate   = tonumber(ARGV[2])
local period = tonumber(ARGV[3])
local cost   = tonumber(ARGV[4])

-- Defensive only: the admin boundary (constraint 13) rejects these before they
-- reach Redis. Guarding here keeps the script total no matter the caller.
if burst < 1 or rate <= 0 or period <= 0 or cost < 1 then
  return { 0, 0, 0, 0 }
end

-- Fractional interval keeps precision for rates that don't divide the period
-- evenly (e.g. 1000 req/hour -> 3.6ms per token) instead of truncating.
local interval = (period * 1000) / rate

local t = redis.call('TIME')
local now = (t[1] * 1000) + math.floor(t[2] / 1000)

local tat = tonumber(redis.call('GET', KEYS[1]) or '0')
-- A missing key, or one whose refill finished long ago, has a TAT in the
-- past: the bucket is full, so pin the TAT to now rather than carrying a
-- stale timestamp forward.
if tat < now then
  tat = now
end

local available = burst - ((tat - now) / interval)

if available < cost then
  -- Deny without mutating state. The wait until `cost` tokens exist is
  -- (cost - available) intervals, measured from now.
  local remaining = available
  if remaining < 0 then
    remaining = 0
  end
  local reset_at = now + ((cost - available) * interval)
  if reset_at < now then
    reset_at = now
  end
  return { 0, math.floor(remaining), math.floor(reset_at), now }
end

-- Allow: push the TAT forward by the cost of this request, atomically.
local new_tat = tat + (cost * interval)
redis.call('SET', KEYS[1], tostring(new_tat))

local remaining = burst - ((new_tat - now) / interval)
if remaining < 0 then
  remaining = 0
end

return { 1, math.floor(remaining), math.floor(new_tat), now }

File: redis/gcra.go
// Package redisclient is the small shared Redis layer both services use: the
// GCRA Lua script (embedded, so rate-limiter and control-plane can never drift
// from a second copy) and the client factory (Sentinel failover by default,
// plain single-endpoint for managed providers via REDIS_URL).
//
// AGENTS.md permits a small shared redisclient/types package; this is it. The
// Lua script and its reply contract live in exactly one place.
package redisclient

import (
	"context"
	_ "embed"
	"fmt"

	"github.com/redis/go-redis/v9"
)

//go:embed gcra.lua
var gcraLua string

// FailoverConfig is the subset of Sentinel connection options the services
// read from the environment. Sentinel-aware only — never a static master
// hostname (constraint 3).
type FailoverConfig struct {
	MasterName    string
	SentinelAddrs []string
	Password      string
}

// NewFailoverClient builds a Sentinel-aware client pointing at whatever node
// currently holds the master.
func NewFailoverClient(cfg FailoverConfig) *redis.Client {
	return redis.NewFailoverClient(&redis.FailoverOptions{
		MasterName:    cfg.MasterName,
		SentinelAddrs: cfg.SentinelAddrs,
		Password:      cfg.Password,
	})
}

// ClientConfig is the union of connection options both services read from the
// environment. Sentinel is the default topology (constraint 3); URL is the
// explicit opt-in for managed single-endpoint Redis (Render Key Value,
// Upstash, ElastiCache) where there is no Sentinel to query. A non-empty URL
// wins over SentinelAddrs.
type ClientConfig struct {
	MasterName    string
	SentinelAddrs []string
	Password      string
	URL           string
}

// NewClient builds the right go-redis client for the config: a Sentinel-aware
// failover client by default, or a plain client for an explicit REDIS_URL.
// Managed Redis providers don't speak the Sentinel protocol, so portability
// to them (Render's Key Value in the demo deployment) is a config choice, not
// a code change. Sentinel remains the primary topology; REDIS_URL is the
// deviation flagged in AGENTS.md constraint 3.
func NewClient(cfg ClientConfig) (*redis.Client, error) {
	if cfg.URL != "" {
		opt, err := redis.ParseURL(cfg.URL)
		if err != nil {
			return nil, fmt.Errorf("parse REDIS_URL: %w", err)
		}
		return redis.NewClient(opt), nil
	}
	return redis.NewFailoverClient(&redis.FailoverOptions{
		MasterName:    cfg.MasterName,
		SentinelAddrs: cfg.SentinelAddrs,
		Password:      cfg.Password,
	}), nil
}

// Params mirrors the script's ARGV contract (constraint 13: the service
// validates before this is reached; the script still guards defensively).
type Params struct {
	Burst     int64 // bucket capacity in tokens, >= 1
	Rate      int64 // tokens per period, > 0
	PeriodSec int64 // seconds per rate period, > 0
	Cost      int64 // tokens consumed by this request, >= 1
}

// CheckResult is the decoded {allowed, remaining, reset_at_ms, now_ms} reply.
// NowMS lets the service compute Retry-After on the same clock the bucket was
// evaluated, avoiding host/Redis clock skew.
type CheckResult struct {
	Allowed   bool
	Remaining int64
	ResetAtMS int64
	NowMS     int64
}

// GCRA wraps the embedded script. One instance per service is fine — the
// script is already loaded and cached by go-redis after the first run.
type GCRA struct {
	script *redis.Script
}

// NewGCRA returns a GCRA wired to the given client.
func NewGCRA(c redis.Cmdable) *GCRA {
	return &GCRA{script: redis.NewScript(gcraLua)}
}

// Check runs one atomic EVAL. It is the only place the script reply is
// decoded, so a contract change touches one file plus the script itself.
func (g *GCRA) Check(ctx context.Context, c redis.Cmdable, key string, p Params) (CheckResult, error) {
	raw, err := g.script.Run(ctx, c, []string{key}, p.Burst, p.Rate, p.PeriodSec, p.Cost).Result()
	if err != nil {
		return CheckResult{}, err
	}
	items, ok := raw.([]interface{})
	if !ok || len(items) != 4 {
		return CheckResult{}, fmt.Errorf("unexpected GCRA reply %#v", raw)
	}
	allowed := items[0].(int64)
	remaining := items[1].(int64)
	resetAt := items[2].(int64)
	now := items[3].(int64)
	// The script's defensive guard returns now=0; by then the service already
	// validated parameters (constraint 13), so this is a programming error and
	// must not surface as a weird-but-successful response.
	if now <= 0 {
		return CheckResult{}, fmt.Errorf("GCRA rejected parameters (now=%d)", now)
	}
	return CheckResult{
		Allowed:   allowed == 1,
		Remaining: remaining,
		ResetAtMS: resetAt,
		NowMS:     now,
	}, nil
}
